test/unittest/StatisticsBackend/CallingUserListenersTests.cpp

```diff
@@ -1448,6 +1457,671 @@ GTEST_INSTANTIATE_TEST_MACRO(
+                discovered_entity_id_));
+    ASSERT_TRUE(writer2);
+    ASSERT_TRUE(writer2->active);
+    ASSERT_EQ(participant->id, writer2->participant->id);
+
+    // Check that the topic is OK
+    ASSERT_TRUE(writer2->topic);
+    ASSERT_EQ(topic2.get(), writer2->topic.get());
+    ASSERT_TRUE(topic2->active);
+    ASSERT_EQ(1, topic2->data_readers.size());
+    ASSERT_EQ(reader2.get(), topic2->data_readers.find(reader2->id)->second.get());
+    ASSERT_EQ(1, topic2->data_writers.size());
+    ASSERT_EQ(writer2.get(), topic2->data_writers.find(writer2->id)->second.get());
+
+    // Check that the locator is OK
+    ASSERT_EQ(1, writer2->locators.size());
+    ASSERT_EQ(rlocator.get(), writer2->locators.begin()->second.get());
+    ASSERT_TRUE(rlocator->active);
+    ASSERT_EQ(1, rlocator->data_readers.size());
+    ASSERT_EQ(reader.get(), rlocator->data_readers.find(reader->id)->second.get());
+    ASSERT_EQ(1, rlocator->data_writers.size());
+    ASSERT_EQ(writer2.get(), rlocator->data_writers.find(writer2->id)->second.get());
+
+    /* Remove the DATAWRITER and DATAREADER on the first TOPIC */
+    DomainEntityDiscoveryArgs datawriter_undiscovery_args([&](
+                EntityId domain_id,
+                EntityId /*entity_id*/,
+                const DomainListener::Status& status)
+            {
+                EXPECT_EQ(monitor_id_, domain_id);
+                EXPECT_EQ(2, status.total_count);
+                EXPECT_EQ(0, status.total_count_change);
+                EXPECT_EQ(1, status.current_count);
+                EXPECT_EQ(-1, status.current_count_change);
+            });
+
+    EXPECT_CALL(domain_listener_, on_datawriter_discovery(monitor_id_, _, _)).Times(1)
+            .WillOnce(Invoke(&datawriter_undiscovery_args, &DomainEntityDiscoveryArgs::on_discovery));
+
+    eprosima::fastrtps::rtps::WriterDiscoveryInfo writer_undiscovery_info (writer_data);
+    writer_undiscovery_info.status = eprosima::fastrtps::rtps::WriterDiscoveryInfo::REMOVED_WRITER;
+
+    // Execution: Call the listener
+    participant_listener_->on_publisher_discovery(participant_, std::move(writer_undiscovery_info));
+    details::StatisticsBackendData::get_instance()->entity_queue_->flush();
+
+    ASSERT_FALSE(writer->active);
+    ASSERT_TRUE(reader->active);
+    ASSERT_TRUE(topic->active);
+    ASSERT_EQ(1, topic->data_readers.size());
+    ASSERT_EQ(reader.get(), topic->data_readers.find(reader->id)->second.get());
+    ASSERT_EQ(1, topic->data_writers.size());
+    ASSERT_EQ(writer.get(), topic->data_writers.find(writer->id)->second.get());
+    ASSERT_TRUE(wlocator->active);
+    ASSERT_TRUE(rlocator->active);
+    ASSERT_TRUE(participant->active);
+    ASSERT_EQ(2, participant->data_readers.size());
+    ASSERT_EQ(reader.get(), participant->data_readers.find(reader->id)->second.get());
+    ASSERT_EQ(2, participant->data_writers.size());
+    ASSERT_EQ(writer.get(), participant->data_writers.find(writer->id)->second.get());
```

Comment:
These are difficult to generalize, I think we would end up adding all the possible casuistry to the input arguments to whatever function we create, and it will only mess the readability of the test.